src/metabase/upload.clj

```diff
@@ -306,24 +321,18 @@
   "Loads a table from a CSV file. If the table already exists, it will throw an error.
    Returns the file size, number of rows, and number of columns."
   [driver db-id table-name ^File csv-file]
-  (let [{col-to-insert->upload-type :extant-columns
-         gen-col->upload-type       :generated-columns} (detect-schema csv-file)
-        cols->upload-type       (merge gen-col->upload-type col-to-insert->upload-type)
+  (let [{:keys [extant-columns generated-columns parse-rows]} (detect-schema csv-file)
+        cols->upload-type       (merge generated-columns extant-columns)
         col-to-create->col-spec (upload-type->col-specs driver cols->upload-type)
-        csv-col-names           (keys col-to-insert->upload-type)]
+        csv-col-names           (keys extant-columns)]
     (driver/create-table! driver db-id table-name col-to-create->col-spec)
-    (try
-      (with-open [reader (io/reader csv-file)]
-        (let [rows (->> (csv/read-csv reader)
-                        (drop 1) ; drop header
-                        (parse-rows col-to-insert->upload-type))]
-          (driver/insert-into! driver db-id table-name csv-col-names rows)
-          {:num-rows          (count rows)
-           :num-columns       (count csv-col-names)
-           :generated-columns (- (count col-to-create->col-spec)
-                                 (count col-to-insert->upload-type))
-           :size-mb           (/ (.length csv-file)
-                                 1048576.0)}))
-      (catch Throwable e
-        (driver/drop-table! driver db-id table-name)
-        (throw (ex-info (ex-message e) {:status-code 400}))))))
+    (with-open [reader (io/reader csv-file)]
+      (let [rows (->> (csv/read-csv reader)
+                      (drop 1) ; drop header
+                      parse-rows)]
```

Comment:
The main change here is `parse-rows` is now a closure that is returned by `detect-schema`, and will remove any columns from the rows that have a column name that matches `_mb_row_id`.